core/src/serve_repair.rs

```diff
@@ -659,6 +667,60 @@ impl ServeRepair {
         true
     }
 
+    fn check_ping_cache(
+        ping_cache: &mut PingCache,
+        request: &RepairProtocol,
+        from_addr: &SocketAddr,
+        identity_keypair: &Keypair,
+    ) -> (bool, Option<Packet>) {
+        let mut rng = rand::thread_rng();
+        let mut pingf = move || Ping::new_rand(&mut rng, identity_keypair).ok();
+        let (check, ping) =
+            ping_cache.check(Instant::now(), (*request.sender(), *from_addr), &mut pingf);
+        let ping_pkt = if let Some(ping) = ping {
+            match request {
+                RepairProtocol::LegacyWindowIndex(_, _, _)
+                | RepairProtocol::LegacyHighestWindowIndex(_, _, _)
+                | RepairProtocol::LegacyOrphan(_, _)
+                | RepairProtocol::LegacyWindowIndexWithNonce(_, _, _, _)
+                | RepairProtocol::LegacyHighestWindowIndexWithNonce(_, _, _, _)
+                | RepairProtocol::LegacyOrphanWithNonce(_, _, _)
+                | RepairProtocol::LegacyAncestorHashes(_, _, _)
+                | RepairProtocol::Pong(_) => None,
+                RepairProtocol::WindowIndex { .. }
+                | RepairProtocol::HighestWindowIndex { .. }
+                | RepairProtocol::Orphan { .. } => {
+                    let ping = RepairResponse::Ping(ping);
+                    Packet::from_data(Some(from_addr), ping).ok()
+                }
+                RepairProtocol::AncestorHashes { .. } => {
+                    let ping = AncestorHashesResponse::Ping(ping);
+                    Packet::from_data(Some(from_addr), ping).ok()
+                }
+            }
+        } else {
+            None
+        };
+        (check, ping_pkt)
+    }
+
+    fn uses_ping_cache_check(request: &RepairProtocol) -> bool {
+        match request {
+            RepairProtocol::LegacyWindowIndex(_, _, _)
+            | RepairProtocol::LegacyHighestWindowIndex(_, _, _)
+            | RepairProtocol::LegacyOrphan(_, _)
+            | RepairProtocol::LegacyWindowIndexWithNonce(_, _, _, _)
+            | RepairProtocol::LegacyHighestWindowIndexWithNonce(_, _, _, _)
+            | RepairProtocol::LegacyOrphanWithNonce(_, _, _)
+            | RepairProtocol::LegacyAncestorHashes(_, _, _)
```

Comment:
Legacy requests should be removed when ping/pong verification is required. I don't see the need to extend it to legacy types now.

Reply:
What is the downside of requiring ping/pong for legacy requests as well?

I think it would be much simpler if we don't tie ping/pong and signed requests together, or add a dependency between the two. These two seem to me separate independent things.

Reply:
changed to apply to legacy requests